### PR TITLE
net, wgengine/capture: encode NAT addresses in pcap stream

### PR DESCRIPTION
**Pull request by twitchyliquid64**

### CODE REVIEW

File: net/packet/packet.go [hunk 1]
```diff
@@ -58,6 +68,11 @@ type Parsed struct {
 	Dst netip.AddrPort
 	// TCPFlags is the packet's TCP flag bits. Valid iff IPProto == TCP.
 	TCPFlags TCPFlag
+
+	// CaptureMeta contains metadata that is used when debugging. It is
+	// usually not present on on a Parsed structure unless debugging is
+	// enabled.
+	CaptureMeta *CaptureMeta
```
Comment (maisem) [inline net/packet/packet.go:75]: drop the `*`, we can just have it be present always.

File: net/packet/packet.go [hunk 2]
```diff
@@ -84,6 +99,9 @@ func (p *Parsed) String() string {
 // and shouldn't need any memory allocation.
 func (q *Parsed) Decode(b []byte) {
 	q.b = b
+	if q.CaptureMeta != nil {
+		*q.CaptureMeta = CaptureMeta{} // Clear any capture metadata if it exists.
```
Comment (maisem) [inline net/packet/packet.go:103]: and always clear

Comment (twitchyliquid64) [reply net/packet/packet.go]: Done

File: net/packet/packet.go [hunk 3]
```diff
@@ -443,11 +461,19 @@ func (q *Parsed) IsEchoResponse() bool {
 // UpdateSrcAddr updates the source address in the packet buffer (e.g. during
 // SNAT). It also updates the checksum. Currently (2022-12-10) only TCP/UDP/ICMP
 // over IPv4 is supported. It panics if called with IPv6 addr.
-func (q *Parsed) UpdateSrcAddr(src netip.Addr) {
+func (q *Parsed) UpdateSrcAddr(src netip.Addr, captureMetadata bool) {
 	if q.IPVersion != 4 || src.Is6() {
 		panic("UpdateSrcAddr: only IPv4 is supported")
 	}
 
+	if captureMetadata {
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply net/packet/packet.go]: Done

File: net/packet/packet.go [hunk 4]
```diff
@@ -34,6 +34,14 @@ const (
 	TCPECNBits TCPFlag = TCPECNEcho | TCPCWR
 )
 
+// CaptureMeta contains metadata that is used when debugging.
+type CaptureMeta struct {
+	SNAT        bool           // SNAT was performed & the address was updated.
```
Comment (maisem) [inline net/packet/packet.go:39]: DidSNAT
DidDNAT

**Review by maisem — APPROVED** [review]

**Pull request merged**